crates/circuit/src/bit.rs

```diff
@@ -0,0 +1,588 @@
+    }
+
+    pub fn extra_info(&self) -> Option<&BitExtraInfo> {
+        match self {
+            BitInfo::Owned { extra, .. } => extra.as_ref(),
+            BitInfo::Anonymous { extra, .. } => extra.as_ref(),
+        }
+    }
+
+    pub(crate) fn register(&self) -> Option<RegisterInfo> {
+        match self {
+            BitInfo::Owned { register, .. } => Some(RegisterInfo::Owning(register.clone())),
+            BitInfo::Anonymous { .. } => None,
+        }
+    }
+}
+
+macro_rules! create_bit_object {
+    ($name:ident, $extra:ty, $extra_exp:expr, $reg:tt) => {
+        #[derive(Debug, Clone, PartialEq, Eq, Hash)]
+        #[doc = concat!("Creates an instance of [", stringify!($name), "].")]
+        pub struct $name(pub(crate) BitInfo);
+
+        impl $name {
+            #[doc = concat!("Creates an anonymous instance of [", stringify!($name), "].")]
+            pub fn new_anonymous() -> Self {
+                Self(BitInfo::new_anonymous(Some($extra_exp)))
+            }
+
+            #[doc = concat!("Returns a reference to the owning register of the [", stringify!($name), "] if any exists.")]
+            pub fn register(&self) -> Option<$reg> {
+                self.0.register().map(|reg| $reg(reg.into()))
+            }
+
+            #[doc = concat!("Returns the index of the [", stringify!($name), "] within the owning register if any exists.")]
+            pub fn index(&self) -> Option<u32> {
+                match &self.0 {
+                    BitInfo::Owned { index, .. } => Some(*index),
+                    _ => None,
+                }
+            }
+        }
+    };
+}
+
+create_bit_object! {ShareableQubit, BitExtraInfo, BitExtraInfo::Qubit{is_ancilla: false}, QuantumRegister}
+
+impl ShareableQubit {
+    /// Check if the Qubit instance is ancillary.
+    pub fn is_ancilla(&self) -> bool {
+        match self.0.extra_info() {
+            Some(BitExtraInfo::Qubit { is_ancilla }) => *is_ancilla,
+            _ => false,
+        }
+    }
+
+    /// Creates an instance of ancilla qubit.
+    pub fn new_anonymous_ancilla() -> Self {
+        Self(BitInfo::new_anonymous(Some(BitExtraInfo::Qubit {
+            is_ancilla: true,
+        })))
+    }
+}
+
+create_bit_object! {ShareableClbit, (), BitExtraInfo::Clbit(), ClassicalRegister}
+
+impl<'py> IntoPyObject<'py> for ShareableQubit {
+    type Target = PyQubit;
+    type Output = Bound<'py, PyQubit>;
+    type Error = PyErr;
+
+    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error> {
+        let ancilla = self.is_ancilla();
+        let bit = PyBit(self.0.clone());
+        let base = PyQubit(self);
+        match ancilla {
+            true => Ok(Bound::new(
+                py,
+                PyClassInitializer::from(bit)
+                    .add_subclass(base.clone())
+                    .add_subclass(PyAncillaQubit(base.clone())),
+            )?
+            .into_super()),
+            false => Bound::new(py, (base, bit)),
+        }
```

Comment:
The variable names read backwards to me and at first I was confused by the inheritance chain being built here. But then I realized bit was the base bit type and base was the qubit object. I think this would be a lot clearer if you did:

```suggestion
        let py_base_bit = PyBit(self.0.clone());
        let py_qubit = PyQubit(self);
        match ancilla {
            true => Ok(Bound::new(
                py,
                PyClassInitializer::from(py_base_bit)
                    .add_subclass(py_qubit.clone())
                    .add_subclass(PyAncillaQubit(py_qubit.clone())),
            )?
            .into_super()),
            false => Bound::new(py, (py_qubit, py_base_bit)),
        }
```

Reply:
Obsolete in 883d82334.